### PR TITLE
Add OCI provider [5/7]

### PR DESCRIPTION
**Pull request by gabriel-samfira**
This adds logic to ensure that a default network exists when we first bootstrap. Also teardown logic.

Depends on: https://github.com/juju/juju/pull/8537

### CODE REVIEW

File: provider/oci/environ.go [hunk 1]
```diff
@@ -57,30 +59,116 @@ func (e *Environ) ecfg() *environConfig {
...
+	request := ociCore.ListInstancesRequest{
+		CompartmentId: compartment,
+	}
+	response, err := e.Compute.ListInstances(context.Background(), request)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	ret := []*ociInstance{}
+	for _, val := range response.Items {
```
Comment (manadart) [inline provider/oci/environ.go]: Worth logging debug messages at the sites of ```continue``` here?

File: provider/oci/networking.go [hunk 2]
```diff
@@ -4,65 +4,981 @@
...
+
+func (e *Environ) getVCNStatus(vcnID *string) (string, error) {
+	request := ociCore.GetVcnRequest{
+		VcnId: vcnID,
+	}
+
+	response, err := e.Networking.GetVcn(context.Background(), request)
+	if err != nil {
+		if e.isNotFound(response.RawResponse) {
+			return "", errors.NotFoundf("vcn not found: %s", *vcnID)
```
Comment (manadart) [inline provider/oci/networking.go]: ```NotFoundf``` suffixes " not found", so you can use simple use something like
```errors.NotFoundf("vcn %q", *vcnID)```

File: provider/oci/environ.go (hunk 1 above)
Comment (gabriel-samfira) [reply provider/oci/environ.go]: Not really. You may have a model that spins up and tares down machines constantly, and as such, you might get a lot of results with a life cycle state of ```terminated```. There is no point in logging them. That line just filters out machines we are not interested in.

File: provider/oci/instance_test.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+func (i *instanceSuite) TestId(c *gc.C) {
+	inst, err := oci.NewInstance(*i.ociInstance, i.env)
+	c.Assert(err, gc.IsNil)
+	id := inst.Id()
+	c.Assert(id, gc.Equals, instance.Id(i.testInstanceID))
+}
+
+func (i *instanceSuite) TestsStatus(c *gc.C) {
```
Comment (manadart) [inline provider/oci/instance_test.go]: Is this a typo?

File: provider/oci/networking.go [hunk 4]
```diff
@@ -4,65 +4,981 @@
...
+		if e.isNotFound(response.RawResponse) {
+			return "", errors.NotFoundf("vcn %q", *vcnID)
+		} else {
+			return "", err
+		}
+	}
+	return string(response.Vcn.LifecycleState), nil
+}
+
+func (e *Environ) getVcn(controllerUUID string) (ociCore.Vcn, error) {
```
Comment (manadart) [inline provider/oci/networking.go]: Probably should be getVCN for consistency with methods above and below.

File: provider/oci/networking.go [hunk 5]
```diff
@@ -4,65 +4,981 @@
...
+
+func (e *Environ) secListName(controllerUUID string) string {
+	return fmt.Sprintf("juju-seclist-%s", controllerUUID)
+}
+
+func (e *Environ) ensureVCN(controllerUUID string) (ociCore.Vcn, error) {
+	logger.Infof("ensuring that VCN is created")
+	if vcn, err := e.getVcn(controllerUUID); err != nil {
+		if !errors.IsNotFound(err) {
+			return ociCore.Vcn{}, err
```
Comment (manadart) [inline provider/oci/networking.go]: errors.Trace(err)

File: provider/oci/networking.go [hunk 6]
```diff
@@ -4,65 +4,981 @@
...
+	}
+	return ociCore.Vcn{}, errors.NotFoundf("no such VCN: %s", *name)
+}
+
+func (e *Environ) secListName(controllerUUID string) string {
+	return fmt.Sprintf("juju-seclist-%s", controllerUUID)
+}
+
+func (e *Environ) ensureVCN(controllerUUID string) (ociCore.Vcn, error) {
+	logger.Infof("ensuring that VCN is created")
```
Comment (jameinel) [inline provider/oci/networking.go]: I'm not sure this Infof is as useful as just reporting on the result of the query. If the query is known to take a long time, having a "preparing/ensuring" is ok, but the more interesting log is almost always the result of the query, and you can combine those into the final output. So possibly drop this to Debug but make sure we have a good message of "VCN not found, creating", or something to that effect.

File: provider/oci/networking_test.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+	request, response := makeGetInstanceRequestResponse(
+		n.testCompartment, n.tags, map[string]string{
+			"id":    n.testInstanceID,
+			"name":  "fake",
+			"az":    "us-phoenix-1",
+			"shape": "VM.Standard1.1",
+		}, ociCore.InstanceLifecycleStateRunning)
+
+	gomock.InOrder(
+		n.compute.EXPECT().GetInstance(context.Background(), request).Return(response, nil).Times(1),
```
Comment (manadart) [inline provider/oci/networking_test.go]: Implicitly times 1.

File: provider/oci/networking.go [hunk 8]
```diff
@@ -4,65 +4,981 @@
...
+	if vcn, err := e.getVcn(controllerUUID); err != nil {
+		if !errors.IsNotFound(err) {
+			return ociCore.Vcn{}, err
+		}
+	} else {
+		return vcn, nil
```
Comment (jameinel) [inline provider/oci/networking.go]: Like probably dropping the above Infof, and just putting one here (maybe at Debug, maybe Info) that gives some details of the VCN that we did find.

File: provider/oci/networking.go [hunk 9]
```diff
@@ -4,65 +4,981 @@
...
+		if e.isNotFound(response.RawResponse) {
+			return "", errors.NotFoundf("seclist not found: %s", *resourceID)
+		} else {
+			return "", err
+		}
+	}
+	return string(response.SecurityList.LifecycleState), nil
+}
+
+func (e *Environ) getSecurityList(controllerUUID string, vcnid *string) (ociCore.SecurityList, error) {
```
Comment (jameinel) [inline provider/oci/networking.go]: got to here so far today, will continue on from here tomorrow.

File: provider/oci/networking.go [hunk 10]
```diff
@@ -4,65 +4,981 @@
...
+
+func (e *Environ) getSeclistStatus(resourceID *string) (string, error) {
+	request := ociCore.GetSecurityListRequest{
+		SecurityListId: resourceID,
+	}
+
+	response, err := e.Firewall.GetSecurityList(context.Background(), request)
+	if err != nil {
+		if e.isNotFound(response.RawResponse) {
+			return "", errors.NotFoundf("seclist not found: %s", *resourceID)
```
Comment (manadart) [inline provider/oci/networking.go]: errors.NotFoundf("seclist: %q", *resourceID)

File: provider/oci/networking.go [hunk 11]
```diff
@@ -4,65 +4,981 @@
...
+		return ociCore.SecurityList{}, errors.NotFoundf("no security lists found for vcn: %v", *vcnid)
+	}
+	for _, val := range response.Items {
+		if tag, ok := val.FreeformTags[tags.JujuController]; ok {
+			if tag == controllerUUID {
+				return val, nil
+			}
+		}
+	}
+	return ociCore.SecurityList{}, errors.NotFoundf("no security lists found for vcn: %v", *vcnid)
```
Comment (manadart) [inline provider/oci/networking.go]: And here.

File: provider/oci/networking.go [hunk 12]
```diff
@@ -4,65 +4,981 @@
...
+	// TODO(gsamfira): maybe "local" would be better?
+	subnetDetails := ociCore.CreateSubnetDetails{
+		AvailabilityDomain: &ad,
+		CidrBlock:          &cidr,
+		CompartmentId:      compartment,
+		VcnId:              vcnID,
+		DisplayName:        &displayName,
+		RouteTableId:       routeRableID,
+		SecurityListIds:    seclists,
+		// DnsLabel:           &providerNetwork.DnsLabel,
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: This is something that we need to decide. This value, together with the hostname set on the vnic and the DNS label of the VCN, will form a FQDN. The DNS label must be unique within the VCN. Allowing the user to set this is not ideal, as he would have to create a sort of map for each availability domain. The number of availability domains may be increased in the future. The user would also have to configure the dns label of the VCN separately.

Generating them automatically may yield in FQDNs that are of no use to anyone (UUIDs?). 

So any thoughts on this are welcome.

File: provider/oci/networking.go [hunk 13]
```diff
@@ -4,65 +4,981 @@
...
+			continue
+		}
+		for _, val := range subnets {
+			cidr := *val.CidrBlock
+			existingCidrBlocks[cidr] = true
+		}
+	}
+
+	if len(missing) > 0 {
+		for ad, _ := range missing {
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: changed

File: provider/oci/networking.go [hunk 14]
```diff
@@ -4,65 +4,981 @@
...
+		return nil, errors.Trace(err)
+	}
+
+	// NOTE(gsamfira): There are some limitations at the moment in regards to
+	// security lists:
+	// * Security lists can only be applied on subnets
+	// * Once subnet is created, you may not attach a new security list to that subnet
+	// * there is no way to apply a security list on an instance/VNIC
+	// * We cannot create a model level security list, unless we create a new subnet for that model
+	// ** that means at least 3 subnets per model, which is something we probably don't want
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: > "You cannot attach a new security list", Is it that you can't change the list (add or remove items), or is it just that you can't change the list to a different list?

It means that given 2 security lists: ```seclist1``` and ```seclist2```, if I create a subnet and assign ```seclist1``` to that subnet, you may not add an aditional ```seclist2``` to that subnet and have its rules applied. You may only change the rules within ```seclist1```.

The problem is, that until OCI supports seclists per VNIC, trying to filter anything is pointless. Opening a port on one machine attached to a subnet, will open the same port on ALL machines attached to the same subnet. There is no way to specify the target IP in the security rule, only the source IP. 

So I did not want to give users a false sense of security. It is far better IMO, to document that we do not filter at the provider level. That way users may take steps to properly set up a firewall, until such time as OCI supports VNIC level security lists.

File: provider/oci/networking.go [hunk 15]
```diff
@@ -4,65 +4,981 @@
...
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, nic := range vnics {
+			if nic.Vnic.SubnetId == nil {
+				continue
+			}
+			if hasSubnetList {
+				if !subIdSet.Contains(*nic.Vnic.SubnetId) {
+					continue
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: I was under the impression that this function returns the subnets that this instance is attached to. It does not care about the NICs themselves. As long as I have 3 nics, 2 of which belong to the same subnet, either one I choose should yield the same result --> I belong to the SubnetId. Am I mistaken?

File: provider/oci/networking.go [hunk 16]
```diff
@@ -4,65 +4,981 @@
...
+			if hasSubnetList {
+				if !subIdSet.Contains(subnetId) {
+					continue
+				} else {
+					subIdSet.Remove(subnetId)
+				}
+			}
+			if info, err := makeSubnetInfo(subnet); err == nil {
+				results = append(results, info)
+			}
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: The subnet listing has no real index. I can sort the map by ID if you feel this is important.

File: provider/oci/networking.go [hunk 17]
```diff
@@ -4,65 +4,981 @@
...
+		subnet, ok := subnets[*iface.Vnic.SubnetId]
+		if !ok || subnet.CidrBlock == nil {
+			continue
+		}
+		addr := network.NewScopedAddress(*iface.Vnic.PrivateIp, network.ScopeCloudLocal)
+		nic := network.InterfaceInfo{
+			InterfaceName:    fmt.Sprintf("unsupported%d", iface.Idx),
+			DeviceIndex:      iface.Idx,
+			ProviderId:       network.Id(*iface.Vnic.Id),
+			MACAddress:       *iface.Vnic.MacAddress,
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: > Is it always safe to dereference this pointer? Is it possible that the provider would just not return a MAC address?

see line 922 above.

> We'll want to be careful about the "InterfaceName: 'unsupported'". Our existing code uses the InterfaceName to match Provider objects to Machine observed objects:

There is no way to guess the interface name inside the booted system, unless we set ```net.ifnames=0``` in userdata, and reboot once, but there is no way to cleanly do that at the moment.

File: provider/oci/networking_test.go (hunk 7 above)
Comment (gabriel-samfira) [reply provider/oci/networking_test.go]: changed

File: provider/oci/networking_test.go [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+		n.compute.EXPECT().ListVnicAttachments(context.Background(), attachRequest).Return(attachResponse, nil),
+		n.netw.EXPECT().GetVnic(context.Background(), vnicRequest[0]).Return(vnicResponse[0], nil),
+		n.netw.EXPECT().ListVcns(context.Background(), vcnRequest).Return(vcnResponse, nil),
+		n.netw.EXPECT().ListSubnets(context.Background(), subnetRequest).Return(subnetResponse, nil),
+	)
+
+	info, err := n.env.NetworkInterfaces(nil, instance.Id(n.testInstanceID))
+	c.Assert(err, gc.IsNil)
+	c.Check(info, gc.HasLen, 1)
+	c.Assert(info[0].Address, gc.Equals, network.NewScopedAddress("1.1.1.1", network.ScopeCloudLocal))
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking_test.go]: Changed

File: provider/oci/networking_test.go [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+	c.Assert(err, gc.IsNil)
+	c.Check(info, gc.HasLen, 1)
+	c.Assert(info[0].CIDR, gc.Equals, "1.0.0.0/8")
+
+	lookFor = []network.Id{
+		network.Id("notHere"),
+	}
+	_, err = n.env.Subnets(nil, instance.Id(n.testInstanceID), lookFor)
+	c.Check(err, gc.ErrorMatches, "failed to find the following subnet ids:.*notHere.*")
+}
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking_test.go]: > I do think we're missing some test coverage here. Things like how we allocate subnets don't seem to be tested, and how we initialize a VCN.

this will be tested indirectly in the next PR (6/7). Without creating package tests, there is no way to test private methods.

Will look at switching the ControllerUUID with model UUIDs. However, this means we will be creating 3 subnets per model. Currently the cleanup code for the networking bits only runs when the controller gets killed, hence the reason for using only the controller label.

Creating 3 subnets per model, in a scenario that might have dozens of models created, will probably be a bit messy. I have to check if OCI has quotas regarding subnets. as well.

File: provider/oci/networking.go [hunk 20]
```diff
@@ -4,65 +4,981 @@
...
+	response, err := e.Networking.CreateInternetGateway(context.Background(), request)
+	if err != nil {
+		return ociCore.InternetGateway{}, errors.Trace(err)
+	}
+
+	if err := e.waitForResourceStatus(
+		e.getInternetGatewayStatus,
+		response.InternetGateway.Id,
+		string(ociCore.InternetGatewayLifecycleStateAvailable),
+		5*time.Minute); err != nil {
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: Changed

File: provider/oci/environ.go [hunk 21]
```diff
@@ -57,30 +59,116 @@ func (e *Environ) ecfg() *environConfig {
...
+		select {
+		case <-retryTimer.Chan():
+			status, err = statusFunc(resId)
+			if err != nil {
+				return err
+			}
+			if status == desiredStatus {
+				return nil
+			}
+			retryTimer.Reset(2 * time.Second)
```
[the remark this answers is not in the document]
Comment (jameinel) [reply provider/oci/environ.go]: I think this got buried in the list of things to discuss. I am curious to consider what we want to do as a retry mechanism.

File: provider/oci/instance_test.go [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+			"name":      "fakeVnicName",
+			"privateIP": "1.1.1.1",
+			"publicIP":  "2.2.2.2",
+		},
+	})
+
+	gomock.InOrder(
+		i.compute.EXPECT().ListVnicAttachments(context.Background(), attachRequest).Return(attachResponse, nil),
+		i.netw.EXPECT().GetVnic(context.Background(), vnicRequest[0]).Return(vnicResponse[0], nil),
+	)
```
[the remark this answers is not in the document]
Comment (jameinel) [reply provider/oci/instance_test.go]: Yes, I do think it would be easier to maintain and understand the tests.

File: provider/oci/networking.go [hunk 23]
```diff
@@ -4,65 +4,981 @@
...
+	tcpProtocolNumber  = "6"
+	udpProtocolNumber  = "17"
+	icmpProtocolNumber = "1"
+	allProtocols       = "all"
 )
 
+// TODO(gsamfira): Use "local" instead? make configurable?
+var DnsLabelTld = "juju"
+
+func (e *Environ) vcnName(controllerUUID string) *string {
```
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: *sigh* missed this one.

> Also consider that the Credential is associated at the Model level, as a given controller might be managing models for different users w/ different credentials.

indeed. Will update scope to model for all networking resources that make sense.

File: provider/oci/networking.go [hunk 24]
```diff
@@ -4,65 +4,981 @@
...
 	"github.com/juju/juju/network"
+
+	ociCore "github.com/oracle/oci-go-sdk/core"
+)
+
+const (
+	// DefaultAddressSpace is the subnet to use for the default juju VCN
+	// An individual subnet will be created from this class, for each
+	// availability domain.
+	DefaultAddressSpace = "10.0.0.0/16"
```
[the remark this answers is not in the document]
Comment (jameinel) [reply provider/oci/networking.go]: Thanks

File: provider/oci/networking.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jameinel) [reply provider/oci/networking.go]: It feels like either the user is setting what they want the DNS labels to be, because they have meaning that they want to ascribe, or it ends up as just a UUID as you mention. So either it is something that we give the user a way to control, or we just don't set it at all.
We *could* do something like model-uuid by default, but have it configurable?

File: provider/oci/networking.go [hunk 25]
```diff
@@ -4,65 +4,981 @@
...
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	to4 := ip.To4()
+	if to4 == nil {
+		return "", errors.Errorf("invalid IPv4 address: %s", DefaultAddressSpace)
+	}
+
+	for i := 0; i <= 255; i++ {
+		to4[2] = byte(i)
```
[the remark this answers is not in the document]
Comment (jameinel) [reply provider/oci/networking.go]: I think we can start here. Given your concern about every-model ending up with 3 /24 subnets, it does start to sound like we may need a way to be more fine grained, but we can do that when we get the feedback it is useful/necessary.

File: provider/oci/networking.go [hunk 26]
```diff
@@ -4,65 +4,981 @@
...
+					return errors.Errorf(
+						"Found a subnet with a different vcnID. This should not happen. Vcn: %s, subnet: %s", *vcnID, *subnet.VcnId)
+				}
+			}
+			request := ociCore.DeleteSubnetRequest{
+				SubnetId: subnet.Id,
+			}
+			// we may need to wait for resource to be deleted
+			response, err := e.Networking.DeleteSubnet(context.Background(), request)
+			// Should we attempt to delete all subnets and return an array of errors?
```
[the remark this answers is not in the document]
Comment (jameinel) [reply provider/oci/networking.go]: I'm guessing this is about the point that you got to.

File: provider/oci/networking.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (jameinel) [reply provider/oci/networking.go]: You're probably right that it is ok for this function.

File: provider/oci/environ.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/environ.go]: It did. I could use the retry package, but it would not save much code. If anything, it will probably add some. The retry package seems to have been written to get around transient errors.

The point of this is not to retry in case of an error, but to poll for the status of a particular resource, given a function that fetches that resource status.It felt wrong to re-purpose a mechanism that is clearly centered around error handling (judging by the code and the function signature that gets passed in), to check statuses.

I would have to create a wrapper funtion that treats the fact that desired status != actual_status as an error, define a new error type that defines this as a transient error, and also define an ```IsFatalError()``` function that ignores it. So it just feels wrong in this scenario.

If the polling interval is too short, I can definitely change it. I don't think there is a need for a complicated back-off timer here, especially considering that in most places, there is a maximum timeout of 5 minutes, after which it will fail.

File: provider/oci/networking.go [hunk 27]
```diff
@@ -4,65 +4,983 @@
...
+
+func (e *Environ) getInternetGatewayStatus(resourceID *string) (string, error) {
+	request := ociCore.GetInternetGatewayRequest{
+		IgId: resourceID,
+	}
+
+	response, err := e.Networking.GetInternetGateway(context.Background(), request)
+	if err != nil {
+		if e.isNotFound(response.RawResponse) {
+			return "", errors.NotFoundf("internet gateway not found: %s", *resourceID)
```
Comment (manadart) [inline provider/oci/networking.go]: ```errors.NotFoundf("internet gateway %q", *resourceID)```

File: provider/oci/provider.go [hunk 28]
```diff
@@ -261,6 +275,16 @@ func (e *EnvironProvider) Open(params environs.OpenParams) (environs.Environ, er
 		return nil, errors.New("compartment-id may not be empty")
 	}
 
+	addressSpace := cfg.addressSpace()
+	if _, ipNET, err := net.ParseCIDR(*addressSpace); err == nil {
+		size, _ := ipNET.Mask.Size()
```
Comment (jameinel) [inline provider/oci/provider.go]: I think the error is probably confusing for users.
A subnet of size "/8" is *larger* than a subnet of size "/16".
If I read it correctly, ipnet.Size() returns the "number of leading ones", so it is the /X in the CIDR.
Are you wanting to say "I must have at least a /16, which means a /8 is ok", or are you saying "you can't include a larger subnet than /16" ?
Maybe if we rephrase it to:
"configured subnet %s/%d is not large enough. please configure a subnet in the range of /8 to /16"
?

Suggestions welcome, I just know that the first time I read over this I was certainly confused.

File: provider/oci/networking_test.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (jameinel) [reply provider/oci/networking_test.go]: SGTM

File: provider/oci/provider.go (hunk 28 above)
Comment (gabriel-samfira) [reply provider/oci/provider.go]: Changed. I have no strong opinions on the phrasing of the error. Whatever you think is less confusing, is fine with me.

File: provider/oci/networking.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: Sorted, see bellow.

File: provider/oci/networking.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (gabriel-samfira) [reply provider/oci/networking.go]: changed

**Comment by jameinel** [comment]
$$merge$$

**Comment by jameinel** [comment]
This looks like an intermittent failure:
       panic: test timed out after 25m0s

goroutine 1241526 [running]:
...
github.com/juju/juju/provider/common_test.(*interruptOnDial).Addresses(0xc4206a0340, 0xd181e0, 0xc42067a2f8, 0x0, 0x0, 0x0, 0x0, 0x0)
	/workspace/src/github.com/juju/juju/provider/common/bootstrap_test.go:636 +0xb7
github.com/juju/juju/provider/common.WaitSSH(0xd17c00, 0xc4201872d0, 0xc42010c8a0, 0xd1bae0, 0x1100ed8, 0x0, 0x0, 0xd1ff00, 0xc4206a0340, 0xd181e0, ...)
	/workspace/src/github.com/juju/juju/provider/common/bootstrap.go:707 +0x4e4
github.com/juju/juju/provider/common_test.(*BootstrapSuite).TestWaitSSHKilledWaitingForDial(0xc4200f48f0, 0xc4204ca780)
	/workspace/src/github.com/juju/juju/provider/common/bootstrap_test.go:646 +0x17f
reflect.Value.call(0xc59760, 0xc4200f48f0, 0x5613, 0xc70ecb, 0x4, 0xc4200a8f70, 0x1, 0x1, 0x1103100, 0xc62720, ...)
	/snap/go/2130/src/reflect/value.go:447 +0x969
reflect.Value.Call(0xc59760, 0xc4200f48f0, 0x5613, 0xc42030ef70, 0x1, 0x1, 0xc4204ca870, 0xc42030ef07, 0x1)
	/snap/go/2130/src/reflect/value.go:308 +0xa4
gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1(0xc4204ca780)
	/workspace/src/gopkg.in/check.v1/check.go:772 +0x575
gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1(0xc4202a7180, 0xc4204ca780, 0xc4206a02e0)
	/workspace/src/gopkg.in/check.v1/check.go:666 +0x7c
created by gopkg.in/check%2ev1.(*suiteRunner).forkCall
	/workspace/src/gopkg.in/check.v1/check.go:663 +0x20c


Running it on the branch manually doesn't show the same failure.

$$merge$$
